client/scripts/controllers/chain/ethereum/compound/api.ts

```diff
@@ -1,48 +1,113 @@
+  public get Contract() { return this._Contract; }
+  public readonly Provider: Web3Provider;
+
+  private _Token: MPond | ERC20VotesComp | undefined;
   public get Token() { return this._Token; }
 
-  // private _Timelock: Timelock;
-  // public get Timelock() { return this._Timelock; }
+  private _isAlpha: boolean;
+  private _isTokenMPond: boolean;
+  public isGovAlpha(c: GovernorAlpha | GovernorCompatibilityBravo): c is GovernorAlpha {
+    return this._isAlpha;
+  }
+  public isTokenMPond(t: MPond | ERC20VotesComp | undefined): t is MPond {
+    return this._isTokenMPond;
+  }
+
+  constructor(
+    _factory: any,
+    contractAddress: string,
+    web3Provider: ExternalProvider
+  ) {
+    this.contractAddress = contractAddress;
+    this.Provider = new Web3Provider(web3Provider);
+    // 12s minute polling interval (default is 4s)
+    this.Provider.pollingInterval = 12000;
+  }
 
   public async init(tokenName: string) {
-    await super.init();
+    // only Alpha has a guardian -- use to distinguish between types of governance contract
+    try {
+      this._Contract = GovernorAlpha__factory.connect(this.contractAddress, this.Provider);
+      await this.Contract.guardian();
+      this._isAlpha = true;
+      console.log(`Found GovAlpha contract at ${this.Contract.address}`);
+    } catch (e) {
+      this._isAlpha = false;
+      console.log(`Found non-GovAlpha Compound contract at ${this.Contract.address}, using GovernorCompatibilityBravo`);
+      this._Contract = GovernorCompatibilityBravo__factory.connect(this.contractAddress, this.Provider);
+    }
+    // TODO: use `quorumVotes` to distinguish between raw IGovernor and BravoCompat, for later support
+
+    // fetch token and derive token type
     // i.e. "uni" or "MPond" -- should refer to a call
-    if (tokenName) {
+    let tokenAddress: string;
+
+    // `.token()` is the way to fetch the token address on GovBravo/related contracts,
+    // but since we have no guarantee we're on Bravo, we do it manually via the below ABI hack
+    if (!tokenName) {
+      tokenName = 'token';
+    }
+
+    // attempt to query token via a provided function call name
+    try {
+      // ABI hack to call arbitrarily named functions on GovAlpha contracts
+      const ABI = [
+        {
+          'inputs': [],
+          'name': tokenName,
+          'outputs': [
+            {
+              'name': '',
+              'type': 'address'
+            }
+          ],
+          'stateMutability': 'view',
+          'type': 'function'
+        }
+      ];
+      const iface = new utils.Interface(JSON.stringify(ABI));
+      const data = iface.encodeFunctionData(tokenName);
+      const resultData = await this.Contract.provider.call({ to: this.Contract.address, data });
+      tokenAddress = utils.getAddress(Buffer.from(utils.stripZeros(resultData)).toString('hex'));
+    } catch (err) {
+      console.error(`Could not fetch token ${tokenName}: ${err.message}`);
```

Comment:
Should we be returning an error here instead of logging an error?

Reply:
I see, this is just part of the attempt to see if a contract is alpha vs bravo; and thus not a 40X error?

Reply:
Hold up -- this isn't an HTTP error at all. This is all taking place on the client/browser.

Reply:
Also it's unrelated to alpha vs bravo. Both have their own voting tokens. The question is whether we're able to access the token (we might not be able to if it's alpha + misconfigured in the database).